src/Gaffer/MetadataAlgo.cpp

```diff
@@ -222,6 +227,86 @@ void bookmarks( const Node *node, std::vector<NodePtr> &bookmarks )
 	}
 }
 
+IECore::InternedString numericBookmarkMetadataName( int bookmark )
+{
+	return g_numericBookmarkBaseName.string() + std::to_string( bookmark );
+}
+
+void setNumericBookmark( ScriptNode *scriptNode, int bookmark, Node *node )
+{
+	if( bookmark < 1 || bookmark > 9 )
+	{
+		throw IECore::Exception( "Values for numeric bookmarks need to be in {1, ..., 9}." );
+	}
+
+	// No other node can have the same bookmark value assigned
+	IECore::InternedString metadataName{ numericBookmarkMetadataName( bookmark ) };
+	for( Node *nodeWithMetadata : Metadata::nodesWithMetadata( scriptNode, metadataName, /* instanceOnly = */ true ) )
+	{
+		Metadata::deregisterValue( nodeWithMetadata, metadataName );
+	}
+
+	if( !node )
+	{
+		return;
+	}
+
+	// Replace a potentially existing value with the one that is to be assigned.
+	int currentValue = getNumericBookmark( node );
+	if( currentValue )
+	{
+		Metadata::deregisterValue( node, numericBookmarkMetadataName( currentValue) );
+	}
+
+	Metadata::registerValue( node, metadataName, new BoolData( true ), /* persistent = */ false );
+}
+
+Node *getNumericBookmark( ScriptNode *scriptNode, int bookmark )
+{
+	// Return the first one we find. There should only ever be just one matching node.
+	for( Node *nodeWithMetadata : Metadata::nodesWithMetadata( scriptNode, numericBookmarkMetadataName( bookmark ), /* instanceOnly = */ true ) )
+	{
+		return nodeWithMetadata;
+	}
+
+	return nullptr;
+}
+
+int getNumericBookmark( const Node *node )
+{
+	// Return the first one we find. There should only ever be just one matching value.
+	for( int bookmark = 1; bookmark < 10; ++bookmark )
+	{
+		if( Metadata::value<BoolData>( node, numericBookmarkMetadataName( bookmark ) ) )
+		{
+			return bookmark;
+		}
+	}
+
+	return 0;
+}
+
+bool numericBookmarkAffectedByChange( const IECore::InternedString &changedKey )
```

Comment:
Boost::spirit ftw!

Reply:
Haha, nothing. I guess I've just been looking for a nail I could hit with this massive hammer that I've learned about ;) Would be interesting to compare performance for a case like this some time, but I've now resorted to using a smaller hammer.